enclave/src/main.rs

```diff
@@ -0,0 +1,160 @@
+
+    #[error("could not convert the hash to a byte array")]
+    ByteWriteError(#[from] std::io::Error),
+}
+
+fn hash(s: &str) -> Result<Vec<u8>, HashError> {
+    let mut hash: u32 = 0;
+
+    for c in s.chars() {
+	let c_u32 = c as u32;
+	hash += c_u32;
+    }
+
+    let mut wtr = vec![];
+    wtr.write_u32::<NativeEndian>(hash)?;
+
+    Ok(wtr)
+}
+
+#[derive(thiserror::Error, Debug)]
+enum LoadProgramError {
+    #[error("hash error")]
+    HashError(#[from] HashError),
+
+    #[error("libbpf error")]
+    LibbpfError(#[from] libbpf_rs::Error),
+}
+
+fn init_runtimes(map: &mut libbpf_rs::Map) -> Result<(), LoadProgramError> {
+    let key = hash(RUNC_CHILD)?;
+    let val = vec![0, 0, 0, 0];
+
+    map.update(&key, &val, libbpf_rs::MapFlags::empty())?;
+
+    Ok(())
+}
+
+fn load_programs(path_base_ts: path::PathBuf) -> Result<(), LoadProgramError> {
+    let skel_builder = EnclaveSkelBuilder::default();
+    let open_skel = skel_builder.open()?;
+    let mut skel = open_skel.load()?;
+
+    let mut path_map_runtimes = path_base_ts.join("map_runtimes");
+    skel.maps_mut().runtimes().pin(&mut path_map_runtimes)?;
+
+    init_runtimes(skel.maps_mut().runtimes())?;
+    
+    let path_map_containers = path_base_ts.join("map_containers");
+    skel.maps_mut().containers().pin(path_map_containers)?;
+
+    let path_program_fork = path_base_ts.join("prog_fork");
+    skel.progs_mut().sched_process_fork().pin(path_program_fork)?;
+
+    let path_program_clone = path_base_ts.join("prog_clone_audit");
+    skel.progs_mut().clone_audit().pin(path_program_clone)?;
+
+    let path_program_syslog = path_base_ts.join("prog_syslog_audit");
+    skel.progs_mut().syslog_audit().pin(path_program_syslog)?;
+
+    let mut link_fork = skel.progs_mut().sched_process_fork().attach()?;
+    let path_link_fork = path_base_ts.join("link_fork");
+    link_fork.pin(path_link_fork)?;
+
+    let mut link_clone = skel.progs_mut().clone_audit().attach_lsm()?;
+    let path_link_clone = path_base_ts.join("link_clone_audit");
+    link_clone.pin(path_link_clone)?;
+
+    let mut link_syslog = skel.progs_mut().syslog_audit().attach_lsm()?;
+    let path_link_syslog = path_base_ts.join("link_syslog_audit");
+    link_syslog.pin(path_link_syslog)?;
+
+    Ok(())
+}
+
+#[derive(thiserror::Error, Debug)]
+enum CleanupError {
+    #[error("regex compilation error")]
+    RegexError(#[from] regex::Error),
+
+    #[error("I/O error")]
+    IOError(#[from] std::io::Error),
+
+    #[error("could not convert path to string")]
+    PathToStrConvError,
+}
+
+fn cleanup(path_base: path::PathBuf, dirname: &String) -> Result<(), CleanupError> {
+    let rx = regex::Regex::new(format!(r#"{}"#, dirname).as_str())?;
+    let paths = fs::read_dir(path_base)?;
+
+    for entry in paths {
+	let entry = entry?;
+	let path = entry.path();
+	let path_s = path.to_str().ok_or(CleanupError::PathToStrConvError)?;
+
+	if !rx.is_match(path_s) {
+	    fs::remove_dir_all(path)?;
+	}
+    }
+
+    Ok(())
```

Comment:
Same as above with the match

Reply:
Hmm... here I can't do it exactly the same way as above, because:

*`if !rx.is_match([...]' is inside the loop
* I don't want to do `Ok(())` unless we are done with the whole forloop. And the regex match happens once per iteration - I want to return an error only if we fail with removing directories.

Could you please write down how would you rewrite those lines? I don't get the clear picture what you would like to see.

Reply:
Oh wait. I misread this, they are in different scopes. Disregard this! Apologies.